### ISSUE TITLE
CDash v3.0.0 seems not to work, installation instructions unclear, please help

### ISSUE THREAD

**Issue by emmenlau** [Kitware/CDash#1056]
I've tried to update our CDash 2.6.0 to 3.0.0 but my mileage was pretty bad, there have been numerous SQL errors and after a few hours of tries and failures I've decided to drop everything and start over clean.

But I fail to get CDash 3.0.0 running at all. I've followed the instructions from https://github.com/Kitware/CDash/blob/master/docs/install.md basically by the letter on an Ubuntu 18.04 server. There are no errors reported from any of the steps. But when I go to my CDash front page, I can just see it redirects to `viewProjects.php` but does not show any projects or login.

CDash log shows a number of errors, the topmost being:
```
[2020-09-01 11:17:02] development.ERROR: Table 'bda_cdash.banner' doesn't exist
```
where `bda_cdash` is the mysql table name.

My questions:
  * Its not clear why this table is missing or who should have created it. Is that my error, did I overlook something in the docs?
  * The installation instructions do not say when/how the tables are initialised by CDash. Does this happen automatically on first start? Could this be added to the docs?
  * The installation instructions do not say how to create the database, database user, and database permissions. I've googled and searched the CDash source code. And it seems that good settings for `mysql` are `DEFAULT CHARACTER SET utf8 COLLATE utf8_general_ci` and `GRANT ALL PRIVILEGES`. But I'm not sure if this is correct. Could this be added to the docs?
  * There used to be a file `install.php` in CDash 2.6.0 that needed to be called first, with CDash set to `development` mode. Curiously, that file still exists and can be called. However calling it, instead of a setup, I get errors that the settings in `config.php` do not allow a database connection. Is `config.php` still required? The installation instructions do not say to update `config.php` or `config.local.php` anymore. Does that mean `install.php` is deprecated? Could it be removed then? Could `config.php` also be removed if not required anymore?

Basically, there seem to be some fundamental steps missing from the installation instructions, is that possible?

Here are my relevant config settings from `.env`:
```
APP_ENV=development
APP_DEBUG=true
APP_URL=http://cdash.intern.biodataanalysis.de
APP_KEY=<some random characters>

DB_CONNECTION=mysql
DB_DATABASE=bda_cdash
DB_USERNAME=bda_cdash
DB_PASSWORD=<some random characters>
```

**Comment by zackgalbreath** [comment]
Hey Mario,

Sorry to hear that this didn't go well for you.

`install.php` is still the script responsible for initially creating the database tables. Since you have `APP_ENV=development` set in your environment file, I would've expected you to be redirected to this page since your database doesn't already exist. I'm curious to figure out why that didn't happen in your case...

Would you please try manually opening install.php in a web browser and filling out the form?

**Comment by emmenlau** [comment]
Thanks @zackgalbreath for the quick reply!

Sadly, I do not get to any form.

I just performed a completely clean installation again using a git clone of tag `v3.0.0`. I also dropped the database before the re-installation. I have the complete installation scripted so its relatively easy for me to try this over and over, if you need.

When all install steps are completed, I open directly the page http://cdash.intern.biodataanalysis.de/install.php. This shows the aforementioned error, about the login into the database being incorrect. But the credentials `root` and the database name `cdash` are taken from `config.php`, not `.env` (see screenshot below).

![cdash-screenshot-20200901](https://user-images.githubusercontent.com/1098087/91857092-b626bc80-ec67-11ea-9a50-cae329193f92.png)

**Comment by zackgalbreath** [comment]
Thanks for giving this a shot! I'll update the install docs shortly & let you know when they're ready to go.

**Comment by emmenlau** [comment]
Also, in case that helps, I have the following php dependency packages installed as system packages. Maybe they are incompatible or too old compared to what you use?

```
#> apt-show-versions |grep php-
php-bcmath:all/bionic 1:7.2+60ubuntu1 uptodate
php-bz2:all/bionic 1:7.2+60ubuntu1 uptodate
php-cli:all/bionic 1:7.2+60ubuntu1 uptodate
php-cli-prompt:all/bionic 1.0.3+dfsg-1 uptodate
php-common:all/bionic 1:60ubuntu1 uptodate
php-composer-ca-bundle:all/bionic 1.1.0-1 uptodate
php-composer-semver:all/bionic 1.4.2-1 uptodate
php-composer-spdx-licenses:all/bionic 1.3.0-1 uptodate
php-curl:all/bionic 1:7.2+60ubuntu1 uptodate
php-gd:all/bionic 1:7.2+60ubuntu1 uptodate
php-json:all/bionic 1:7.2+60ubuntu1 uptodate
php-json-schema:all/bionic 5.2.6-1 uptodate
php-ldap:all/bionic 1:7.2+60ubuntu1 uptodate
php-mbstring:all/bionic 1:7.2+60ubuntu1 uptodate
php-mysql:all/bionic 1:7.2+60ubuntu1 uptodate
php-psr-log:all/bionic 1.0.2-1 uptodate
php-symfony-console:all/bionic-security 3.4.6+dfsg-1ubuntu0.1 uptodate
php-symfony-debug:all/bionic-security 3.4.6+dfsg-1ubuntu0.1 uptodate
php-symfony-filesystem:all/bionic-security 3.4.6+dfsg-1ubuntu0.1 uptodate
php-symfony-finder:all/bionic-security 3.4.6+dfsg-1ubuntu0.1 uptodate
php-symfony-polyfill-mbstring:all/bionic 1.6.0-2 uptodate
php-symfony-process:all/bionic-security 3.4.6+dfsg-1ubuntu0.1 uptodate
php-xml:all/bionic 1:7.2+60ubuntu1 uptodate
```

**Comment by zackgalbreath** [comment]
Here's my proposed update to the docs:

https://github.com/Kitware/CDash/compare/improve_install_docs?expand=1
https://github.com/Kitware/CDash/blob/67b14ca293fcd5321cfca1e476557bb3360e7c70/docs/install.md

The main change is to make your config settings in `app/cdash/config/config.local.php` (not `.env`). Remember to call `php artisan config:migrate` when you're done.

I don't think this will solve the mystery of why you didn't get automatically redirected to `install.php`, but hopefully it will allow you to run the install script & create the database tables.

Please give it a try when you get a chance & let us know how it goes!

**Comment by emmenlau** [comment]
Ok, now my mileage is a lot better! I can generate the exported config in `.env`, then add an encryption key, and finally call `install.php`. However this still just shows a page with a message that I should check the installation:

![cdash-screenshot-20200901b](https://user-images.githubusercontent.com/1098087/91875015-360a5200-ec7b-11ea-8f7f-e0024cdc6eee.png)

There are no messages in the cdash log, and no errors in Apache logs. Any ideas what I can try next?

**Comment by zackgalbreath** [comment]
Yeah, that still doesn't look quite right. Can you double check that your web server is configured to allow content to be served over symbolic links?

Try opening up this address in your web browser:
http://cdash.intern.biodataanalysis.de/img/cdash.png

**Comment by emmenlau** [comment]
Argh! I think I've made quite a step forward. Is it possible that contrary to previous versions of CDash, in 3.0.0 the database must *not* exist prior to calling `install.php`?

As I wrote above, I assumed the database must exist because the user may not have permissions to create new databases. However after not creating the database prior to calling `install.php`, I've been able to perform an installation.

This is also slightly contrary to what I found in the code. When reading `install.php` there is an explicit mention of `CREATE DATABASE IF NOT EXIST`. In any case, it may be better to accept existing databases because it requires less user privileges, no?

Also, generally CDash 3.0.0 seems still quite a bit rough on the edges, if I may say so :-) During creation of my first project, the images for the "previous" and "next" buttons are broken, they link to an image url that contains `/undefined/`. After project creation, there is a link to the project page. However this link leads to an "page does not exist" error page. Trying another link goes to the dashboard but it shows "no projects". The only way back to the "correct" dashboard is via manually typing the base CDash url, then the new project correctly shows up :-)

**Comment by zackgalbreath** [comment]
Glad to hear you're making progress. When I get a bit of free time I'll try walking through the "install from scratch" workflow again on a base Ubuntu VM to see if there's anything I can do to further improve the docs.

> During creation of my first project, the images for the "previous" and "next" buttons are broken, they link to an image url that contains `/undefined/`

Can you double check that `APP_URL` is set properly in `.env`? This should have happened automatically after running `php artisan config:migrate`.

**Comment by emmenlau** [comment]
Thanks again! I'm making pretty good progress overall, the second project is created and the submission of builds finally works!

Only the above-mentioned problems consistently remain. It seems as soon as I open a project and navigate to more detailed pages, many urls incorrectly start with `project/undefined/` or `project/1/undefined/`, like `http://cdash.intern.biodataanalysis.de/project/1/undefined/img/help.gif`. When I remove this prefix, the images can be seen and the linked pages can be accessed. Many other links work, but especially images are often broken.

I've checked that `APP_URL` is set properly in `.env` so this was taken correctly from `php artisan config:migrate`.

**Comment by zackgalbreath** [comment]
Nice, sounds like you're pretty close. I'll try to reproduce this situation locally, but this is what I would expect to work:

* Make sure `APP_URL` is properly set in `.env`, and that `MIX_APP_URL="${APP_URL}"` appears further down in the file.
* Run `npm run dev` from the root of your CDash installation.

**Comment by emmenlau** [comment]
> * Make sure `APP_URL` is properly set in `.env`, and that `MIX_APP_URL="${APP_URL}"` appears further down in the file.

This was the problem! I was confused about this step: Should I copy `.env.example` to `.env` even when later using `php artisan config:migrate`? I understood these steps to be mutually exclusive.

I performed these steps in order now, first copy then migrate, and it works correctly!!! Hooray!

Maybe this section could be slightly further clarified in the installation docs? Currently the headline reads **Install steps: initial installation only** which made me think that this step is _not_ used on upgrades. A minor improvement may be to call the headline **Install steps: initial installation or first upgrade only**? Just an idea, though!

Thanks a lot, @zackgalbreath !!!

**Comment by zackgalbreath** [comment]
You're welcome. Thanks for sticking with this. I'll update the install docs with your suggestions.
